From a3a8784454692dd72e5d5d34dcdab17b4420e74c Mon Sep 17 00:00:00 2001
From: Sasha Levin <sasha.levin@oracle.com>
Date: Mon, 29 Dec 2014 09:39:01 -0500
Subject: [PATCH] KEYS: close race between key lookup and freeing

Git-commit: a3a8784454692dd72e5d5d34dcdab17b4420e74c
Patch-mainline: v3.19 
References: bnc#912202
Target: SLE-11 SP3

When a key is being garbage collected, it's key->user would get put before
the ->destroy() callback is called, where the key is removed from it's
respective tracking structures.

This leaves a key hanging in a semi-invalid state which leaves a window open
for a different task to try an access key->user. An example is
find_keyring_by_name() which would dereference key->user for a key that is
in the process of being garbage collected (where key->user was freed but
->destroy() wasn't called yet - so it's still present in the linked list).

This would cause either a panic, or corrupt memory.

Fixes CVE-2014-9529.

Backport to v3.0:
There have similar codes in key.c in v3.0, so backported this patch.
commit id: a3a8784454692dd72e5d5d34dcdab17b4420e74c

Signed-off-by: Sasha Levin <sasha.levin@oracle.com>
Signed-off-by: David Howells <dhowells@redhat.com>
Cc: Jiri Slaby <jslaby@suse.cz>
Signed-off-by: Lee, Chun-Yi <jlee@suse.com>
---
 security/keys/gc.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-3.0-SLE11-SP3/security/keys/key.c
===================================================================
--- linux-3.0-SLE11-SP3.orig/security/keys/key.c
+++ linux-3.0-SLE11-SP3/security/keys/key.c
@@ -656,12 +656,12 @@ found_dead_key:
 	if (test_bit(KEY_FLAG_INSTANTIATED, &key->flags))
 		atomic_dec(&key->user->nikeys);
 
-	key_user_put(key->user);
-
 	/* now throw away the key memory */
 	if (key->type->destroy)
 		key->type->destroy(key);
 
+	key_user_put(key->user);
+
 	kfree(key->description);
 
 #ifdef KEY_DEBUGGING
